dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java

```diff
@@ -956,19 +962,18 @@ private void finishFetch()
                 return;
             }
 
-            streamPosOfFetches.add(streamPosOfFetchStart);
-            streamPosOfFetchStart += totalDataInCurrentRequest;
-
             if (this.tokenBin == null || tokenLen > this.tokenBin.length)
             {
                 this.tokenBin = new byte[tokenLen];
             }
-            else
+            dis.readFully(this.tokenBin,0,tokenLen);
+            
+            this.streamPosOfFetchStart += totalDataInCurrentRequest;
+            synchronized (streamPosOfFetches)
```

Comment:
is it possible for "this.tokenBin" to be empty or invalid at this point? Is there any way to confirm the token?

Reply:
If we are given an invalid token at this point it would be a significant issue. I don't think there is a good recovery method in that case. We do have logging if a token is rejected so we would be able to trace the issue.